3p/ampcontext.js

```diff
@@ -62,24 +110,38 @@ export class AmpContext {
    *    every time we receive a page visibility message.
    */
   observePageVisibility(callback) {
-    return this.client_.makeRequest(
-        MessageType.SEND_EMBED_STATE,
-        MessageType.EMBED_STATE,
-        callback);
+    const visibilityCallbacksLength = this.visibilityCallbacks_.length;
+
+    this.visibilityCallbacks_.push(callback);
+
+    return () => {
+      this.visibilityCallbacks_.splice(visibilityCallbacksLength + 1, 1);
+    };
   };
 
   /**
    *  Send message to runtime to start sending intersection messages.
-   *  @param {function(Object)} callback Function to call every time we receive
-   *  an intersection message.
+   *  @param {function(Array<Object>)} callback Function to call every time we
+   *    receive an intersection message.
    *  @returns {function()} that when called stops triggering the callback
    *    every time we receive an intersection message.
    */
   observeIntersection(callback) {
-    return this.client_.makeRequest(
+    const unlisten = this.client_.makeRequest(
         MessageType.SEND_INTERSECTIONS,
         MessageType.INTERSECTION,
-        callback);
+        intersection => {
+          callback(intersection.changes);
+        });
+
+    // Call the callback with the value that was transmitted when the
+    // iframe was drawn. Called in nextTick, so that callers don't
+    // have to specially handle the sync case.
+    nextTick(this.win_, () => {
```

Comment:
let's leave the logic here for now. it will be a breaking change to those who rely on. I'll create a deprecation announcement first.